### PR TITLE
WIP: publish large payloads

### PR DESCRIPTION
**Pull request by bertmelis**
Initial version:
only tested for QOS0
only tested on ESP8266
ESP32 shared resources not protected by semaphore

### CODE REVIEW

**Comment by mknjc** [comment]
Currently the PayloadHandler doesn't know how much of it's returned pointer will be written.
It would be nice to have a way to know how much data is expected to send and also a way to tell: "I've got only x bytes currently. Come again later"

**Comment by bertmelis** [comment]
I've thought about that. But I wanted to avoid an intermediate buffer and you can only return 1 variable. I also wanted to avoid passing the TCP client to the callback.

How would you like to have the handler's signature?
What are you intending to do with the returned size?

The TCP-client works like client.write(uint8* buff, size_t len) and returns the amount written.

**Comment by mknjc** [comment]
One idea I've in my mind is a signature like: `payloadHandler(off_t startIndex, size_t& length)`
The callsite looks something like:
```
size_t lengthToWrite = std::min(_largePayloadLength, _client.space());
std::byte* dataToWrite = PayloadHandler (_largePayloadIndex, lengthToWrite);
if(dataToWrite != nullptr) {
size_t written = _client.add(dataToWrite, lengthToWrite);
_largePayloadIndex += written;
...
}
```
With this the payload handler knows how much byte it has to load from for example a external sram and could also request to send less.

File: src/AsyncMqttClient.cpp [hunk 1]
```diff
@@ -890,6 +908,73 @@ uint16_t AsyncMqttClient::publish(const char* topic, uint8_t qos, bool retain, c
   }
 }
 
+uint16_t AsyncMqttClient::publish(const char* topic, uint8_t qos, bool retain, AsyncMqttClientInternals::PayloadHandler handler, size_t length, bool dup, uint16_t message_id) {
+  if (!_connected) return 0;
```
[the remark this answers is not in the document]
Comment (bertmelis) [reply src/AsyncMqttClient.cpp:912]: yes. This has to be added. In both publish methods.

File: src/AsyncMqttClient.cpp [hunk 2]
```diff
@@ -890,6 +908,73 @@ uint16_t AsyncMqttClient::publish(const char* topic, uint8_t qos, bool retain, c
...
+      packetId = message_id;
+    } else {
+      packetId = _getNextPacketId();
+    }
+
+    packetIdBytes[0] = packetId >> 8;
+    packetIdBytes[1] = packetId & 0xFF;
+  }
+
+  size_t written = _client.add(fixedHeader, 1 + remainingLengthLength);
```
[the remark this answers is not in the document]
Comment (bertmelis) [reply src/AsyncMqttClient.cpp:956]: I left it out but it should be in as there is only one try to send the header.

File: src/AsyncMqttClient/Callbacks.hpp [hunk 3]
```diff
@@ -13,6 +13,7 @@ typedef std::function<void(uint16_t packetId, uint8_t qos)> OnSubscribeUserCallb
 typedef std::function<void(uint16_t packetId)> OnUnsubscribeUserCallback;
 typedef std::function<void(char* topic, char* payload, AsyncMqttClientMessageProperties properties, size_t len, size_t index, size_t total)> OnMessageUserCallback;
 typedef std::function<void(uint16_t packetId)> OnPublishUserCallback;
+typedef std::function<const char*(size_t index)> PayloadHandler;
```
[the remark this answers is not in the document]
Comment (bertmelis) [reply src/AsyncMqttClient/Callbacks.hpp:16]: QOS0 doesn't give a packetId (MQTT spec).

File: examples/LargePayloads/LargePayloads.ino [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+
+void onMqttPublish(uint16_t packetId) {
+  Serial.println("Publish acknowledged.");
+  Serial.print("  packetId: ");
+  Serial.println(packetId);
+}
+
+void connectToWifi() {
+  Serial.println("Connecting to Wi-Fi...");
+  WiFi.begin("WIFI_SSID", "WIFI_PASSWORD");
```
[the remark this answers is not in the document]
Comment (bertmelis) [reply examples/LargePayloads/LargePayloads.ino:54]: Hmm, dirty cutting and lasting from my side...

**Comment by bertmelis** [comment]
I assume you mean lengthToWrite to be the min between space and "remaininglength".

But it's indeed useful to include the size that will be written so data can be fetched.

Will try to make an update tomorrow.

**Comment by bertmelis** [comment]
I should test with QOS1 and QOS2 also shouldn't I?

I do want to commit a few other things also:
- the onPublish callback is only used for QOS1 and QOS2. But here it would be handy also to know when publishing has finished. Otherwise you'd have to rely on the payloadHandler to calculate when the last piece will be written. An option would be to call the onPublish also on QOS0.
- in the (internal) _onData handler _lastServerActivity is only reset when a *new* MQTT message arrives. On large payloads (and short keepalives) like OTA this makes the ESP send a ping. This is unneeded and interferes with the ongoing communication.

**Comment by bertmelis** [comment]
close for now. Maybe one day try to re-implement.

**Pull request closed without merging**